### PR TITLE
Adding bounding box class

### PR DESCRIPTION
**Pull request by shimwell**
As requested in issue #2456 this PR adds a ```BoundingBox``` class and makes use of it on the Python side of the code.

The bounding box class behaves like the previous bounding box tuple of lower left and upper right but adds a couple of useful methods (center and volume). Other useful methods can be added later if people want

Tests have been added to check the volume and center are correct and that the inputs are well formed (correct length and correct types)

### CODE REVIEW

**Comment by pshriwise** [comment]
I'm on board @shimwell. I'll ask that we allow indexing of the `BoundingBox` class for backward compatibility (meaning you'll get the lower corner with `bbox[0]`). 

Also, what do you think about using something like `lower_left` (and `upper_right`) or even `llc` (and `urc`) as names for properties of the box?

**Comment by shimwell** [comment]
Thanks Patrick. I've added the lower_left and upper_right methods suggested. The class inherits from tuple so it should behave in the same manner as before with regards to indexing and bbox[0] and bbox[1] should be the same ✅

File: openmc/bounding_box.py [hunk 1]
```diff
@@ -0,0 +1,54 @@
...
+    upper_right
+        The x, y, z coordinates of the upper right corner of the bounding box
+    volume: float
+        The volume of the bounding box in cm3
+    """
+
+    def __init__(self, corners):
+
+        check_type("corners", corners, tuple)
+        check_type("corners", corners[0], np.ndarray)
```
Comment (pshriwise) [inline openmc/bounding_box.py:29]: Requiring this to be a tuple of NumPy array feels a little restrictive. Creating a bounding box from scratch currently requires a line like this:

```python
bbox = openmc.BoundingBox((np.asarray([0.0, 0.0, 0.0]), np.asarray([1.0, 1.0, 1.0])))
```

whereas with the allowance of general iterables of floats this reduces to:

```python
bbox = openmc.BoundingBox([[0.0, 0.0, 0.0], [1.0, 1.0, 1.0]])
```

which I personally find a little easier to work with.


What would you say to converting the arguments to `np.ndarray` as needed with `np.asarray`?

**Review by pshriwise — CHANGES_REQUESTED** [review]
I like this addition @shimwell. I've considered adding it before, but it felt like overkill...until I started playing around with this branch that is! 

While I do think that inheritance from native Python types is often overlooked as an option, the use of an immutable one here has a couple of complications/restrictions:

As you noted in the PR conversation, you can in fact index the tuple without any additional code to get the lower left and upper right corners of the box. This is because the `__new__` method of the inherited tuple class is also creating a tuple container under the hood. While this isn't necessarily problematic, it's inefficient to store the corners of the bounding box twice. This can be demonstrated by the following code snippet:

```python
llc = np.asarray([0.0, 0.0, 0.0])
urc = np.asarray([1.0, 1.0, 1.0])
bbox = openmc.BoundingBox((llc, urc)
print(id(bbox[:])) # returns the tuple itself
print(id(bbox.corners))
```
This will show two different IDs (i.e. two different tuple instances) associated with this `BoundingBox` instance.

This is more of a preference regarding the mutability of our bounding boxes, but in some of the ray tracing kernels I've worked with there are really handy `BoundingBox.update` methods that allow one to pass a point (or another `BoundingBox` instance) to update the corners of the bounding box. This might be useful in the `Region.bounding_box` method for instance.

The last thing I'll note is that the current implementation of this class is really mutable in the end. The tuple will contain immutable _references_ to the NumPy arrays, but the arrays themselves can still be changed. So something like

```python
bbox.corners[0][1] = 22.0
```
will result in a change to the bounding values.

**Comment by paulromano** [comment]
@pshriwise Note that I sent a PR to @shimwell's fork with a refactor of this branch that should address most of your comments: https://github.com/shimwell/openmc/pull/28

**Comment by pshriwise** [comment]
> @pshriwise Note that I sent a PR to @shimwell's fork with a refactor of this branch that should address most of your comments: [shimwell#28](https://github.com/shimwell/openmc/pull/28)

Oh nice! I'll let that come in and then have another look. Thanks for the notice!

**Comment by shimwell** [comment]
Really sorry for missing Paul's PR. I'm not sure how that one escaped me. Sorry it caused extra reviewing effort but looks like you two spotted the same improvements

**Comment by shimwell** [comment]
If this PR is successful then I have a follow on PR that makes use of this bounding box to automatically determine the universe.plot origin and width. I won't open that PR quite yet in case this bounding box one doesn't go ahead but the branch is [here](https://github.com/shimwell/openmc/tree/auto_origin_and_width_for_plots) so I don't lose it

File: openmc/bounding_box.py [hunk 2]
```diff
@@ -0,0 +1,57 @@
...
+    ----------
+    lower_left : iterable of float
+        The x, y, z coordinates of the lower left corner of the bounding box in [cm]
+    upper_right : iterable of float
+        The x, y, z coordinates of the upper right corner of the bounding box in [cm]
+
+    Attributes
+    ----------
+    center : numpy.ndarray
+        x, y, z coordinates of the center of the bounding box in [cm].
```
Comment (pshriwise) [inline openmc/bounding_box.py:23]:
```suggestion
        x, y, z coordinates of the center of the bounding box in [cm]
```

File: openmc/lib/cell.py [hunk 3]
```diff
@@ -101,7 +103,7 @@ class Cell(_FortranObjectWithID):
         Name of the cell
     num_instances : int
         Number of unique cell instances
-    bounding_box : 2-tuple of numpy.ndarray
+    bounding_box : openmc.BoundingBox
```
Comment (pshriwise) [inline openmc/lib/cell.py:106]: The description of this property in the line below is a little off now.

**Review by pshriwise — CHANGES_REQUESTED** [review]
A couple of small nit-picks here.

One other design question I have is whether or not this class would allow degenerate bounding boxes (where one or more values of the lower-left corner are greater than the upper-right)?

File: openmc/lib/cell.py (hunk 3 above)
Comment (shimwell) [reply openmc/lib/cell.py:106]: thanks Patrick, I have changed the docstring to match the equivalent docstring in cell.py

**Comment by shimwell** [comment]
>  whether or not this class would allow degenerate bounding boxes

Currently the bounding box tuple in the code allows degenerate bounding boxes but we could add some checking to the ```def __new__``` method to see if the lower left values are all smaller than the upper right values.

Happy to add that in here or as a separate PR.

I guess the ```inf``` values will need some special handling and the extra checking will slightly slow down the creation of bounding boxes.

**Comment by pshriwise** [comment]
> I guess the `inf` values will need some special handling and the extra checking will slightly slow down the creation of bounding boxes a tiny amount.

**Comment by pshriwise** [comment]
Sorry for the inadvertent close @shimwell. Wrong button!

**Review by pshriwise — APPROVED** [review]
I think that this implementation is sound as-is and supports what our current llc/urc tuple is capable of. I'll submit some of my suggestions on design in a separate PR an we can discuss them there. Thanks again for this addition!

**Comment by shimwell** [comment]
Thanks for the review Patrick and the nice corrections. I shall leave those additions for you on another PR as it sounds like you have a nice clear idea of what is needed.

**Pull request merged**